search/binary_search.cpp

```diff
@@ -1,56 +1,148 @@
+ *
+ * Worst-case time complexity	O(log n)
+ * Best-case time complexity	O(1)
+ * Average time complexity	    O(log n)
+ * Worst-case space complexity  0(1)
+ *
+ * @author [Lajat Manekar](https://github.com/Lazeeez)
+ * @author Unknown author
+ *******************************************************************************/
+
+#include <algorithm>  /// for std::sort function
+#include <cassert>    /// for std::assert
+#include <iostream>   /// for IO operations
+#include <vector>     /// for std::vector
+/******************************************************************************
+ * @namespace search
+ * @brief Searching algorithms
+ *******************************************************************************/
+namespace search {
+
+/******************************************************************************
+ * @namespace binary_search
+ * @brief Binary search searching algorihm
+ *******************************************************************************/
+namespace binary_search {
+
+/******************************************************************************
+ * @brief The main function which implements binary search
+ * @param arr vector to be searched in
+ * @param val value to be searched
+ * @returns @param int index of val in vector arr
+ *******************************************************************************/
+uint64_t binarySearch(std::vector<uint64_t> arr, uint64_t val) {
+    uint64_t low = 0;                // set the lowest point of the vector.
+    uint64_t high = arr.size() - 1;  // set the highest point of the vector.
+
+    while (low <= high) {
+        uint64_t m = low + (high - low) / 2;  // set the pivot point
+
+        if (val == arr[m]) {
             return m;
-        else if (key < a[m])
-            r = m - 1;
-        else
-            l = m + 1;
+        } /****************************************************
+           * if pivot point is the val, return it,
+           * else check if val is greater or smaller than pivot value
+           * and set the next pivot point accordingly.
+           ****************************************************/
+        else if (val < arr[m]) {
+            high = m - 1;
+        } else {
+            low = m + 1;
+        }
     }
-    return -1;
+    return -1;  // if val is not in the array, return -1.
 }
 
-/** main function */
-int main(int argc, char const* argv[]) {
-    int n, key;
-    std::cout << "Enter size of array: ";
-    std::cin >> n;
-    std::cout << "Enter array elements: ";
+}  // namespace binary_search
 
-    int* a = new int[n];
+}  // namespace search
 
-    // this loop use for store value in Array
-    for (int i = 0; i < n; i++) {
-        std::cin >> a[i];
-    }
+/*******************************************************************************
+ * @brief Self-test implementation #1
+ * @returns void
+ *******************************************************************************/
+static void test1() {
```

Comment:
Let's make this only `void` and create a static `tests` function, which will run all tests. 😄
```suggestion
void test1() {
```